google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,195 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "monitor"
+require "concurrent"
+require "google/cloud/pubsub/publish_result"
+require "google/cloud/pubsub/service"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Publisher
+      #
+      class Publisher
+        include MonitorMixin
+
+        MAX_COUNT = 2000
+        MAX_SIZE = 10000000
+
+        attr_reader :service, :queue
+        attr_reader :max_size, :max_count
+        attr_reader :thread_pool, :publish_task
+
+        def initialize service, max_size: MAX_SIZE, max_count: MAX_COUNT
+          @service = service
+          @max_size = max_size
+          @max_count = max_count
+
+          @queue = {}
+          @thread_pool = Concurrent::FixedThreadPool.new(
+            [2, Concurrent.processor_count * 2].max,
+            fallback_policy: :caller_runs
+          )
+          @publish_task = Concurrent::TimerTask.new(execution_interval: 0.25) do
+            flush
+          end
+
+          # init MonitorMixin
+          super()
+        end
+
+        def publish topic_name, data = nil, attributes = {}, &block
+          topic_name = service.topic_path topic_name
+          msg = create_pubsub_message data, attributes
+
+          synchronize do
+            batch = @queue[topic_name]
+            if batch.nil?
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            elsif !batch.try_add(msg, block, max_count: max_count,
+                                             max_size: max_size)
+              # We hit a threshold, publish existing batch.
+              publish_batch_async topic_name, batch
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            end
+            @queue[topic_name] = batch
+          end
+          nil
+        end
+
+        def start
```

Comment:
FWIW, I don't understand why `#start` and `#stop` are needed either. We only added these methods because they were in the spec. It does add some confusion though, since there are two constructs that can be stopped, the `TimerTask`, and the `ThreadPool`. I'm fine with removing these methods if that is allowed per the spec.